File: src/main/java/ash/java/graphql/fields/GenreSchema.java
package ash.java.graphql.fields;

import ash.java.graphql.data.GenreDao;
import ash.java.graphql.types.genre.GenreType;
import graphql.schema.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;

@Service
public class GenreSchema implements FieldProducer {

    private GenreDao genreDao;

    @Autowired
    public GenreSchema(GenreDao genreDao) {
        this.genreDao = genreDao;
    }

    private GraphQLFieldDefinition genreListFieldDefinition = newFieldDefinition()
            .type(new GraphQLList(new GenreType().getGraphQlType()))
            .name("genres")
            .dataFetcher(env -> genreDao.getAllMovieGenres())
            .build();

    public GraphQLFieldDefinition getFieldDefinition() {
        return genreListFieldDefinition;
    }
}

File: src/main/kotlin/ash/kotlin/graphql/fields/GenreSchema.kt
package ash.kotlin.graphql.fields

import ash.kotlin.graphql.data.GenreDao
import ash.kotlin.graphql.types.genre.GenreType
import graphql.schema.*

class GenreSchema constructor(private val dao: GenreDao) : FieldProducer
{
    override fun getFieldDefinition(): GraphQLFieldDefinition
    {
        return GraphQLFieldDefinition.newFieldDefinition()
                .type(GraphQLList(GenreType().getGraphQlType()))
                .name("genres")
                .dataFetcher { _ -> dao.getAllMovieGenres() }
                .build()
    }
}